doc/rtd/topics/module_creation.rst

```diff
@@ -0,0 +1,94 @@
+.. _module_creation:
+
+Module Creation
+***************
+
+Much of cloud-init functionality is provided by :ref:`modules<modules>`.
+All modules follow a similar layout in order to provide consistent execution
+and documentation. Use the example provided here to create a new module.
+
+Example
+=======
+.. code-block:: python
+
+    # This file is part of cloud-init. See LICENSE file for license information.
+    """Example Module: Shows how to create a module"""
+
+    from logging import Logger
+
+    from cloudinit.cloud import Cloud
+    from cloudinit.config.schema import MetaSchema, get_meta_doc
+    from cloudinit.distros import ALL_DISTROS
+    from cloudinit.settings import PER_INSTANCE
+
+    MODULE_DESCRIPTION = """\
+    Description that will be used in module documentation.
+
+    This will likely take multiple lines.
+    """
+
+    meta: MetaSchema = {
+        "id": "cc_example",
+        "name": "Example Module",
+        "title": "Shows how to create a module",
+        "description": MODULE_DESCRIPTION,
+        "distros": [ALL_DISTROS],
+        "frequency": PER_INSTANCE,
+        "examples": [
+            "cc_example: example1",
+            "cc_example: ['example', 2]",
+        ],
+    }
+
+    __doc__ = get_meta_doc(meta)
+
+
+    def handle(name: str, cfg: dict, cloud: Cloud, log: Logger, args: list):
+        log.debug(f"Hi from module {name}")
+
+
+Guidelines
+==========
+
+* Create a new module in the ``cloudinit/config`` directory with a `cc_`
+  prefix.
+* Your module must include a ``handle`` function. The arguments are:
+
+  * ``name``: The module name specified in the configuration
+  * ``cfg``: A configuration object that is the result of the merging of
+    cloud-config configuration with any datasource provided configuration.
+  * ``cloud``: A cloud object that can be used to access various datasource
+    and paths for the given distro and data provided by the various datasource
+    instance types.
+  * ``log``: A logger object that can be used to log messages.
+  * ``args``: An argument list. This is usually empty and is only populated
+    if the module is called independently from the command line.
+
+* If your module introduces any new cloud-config keys, you must provide a
+  schema definition in `cloud-init-schema.json`_.
+* The ``meta`` variable must exist and be of type `MetaSchema`_.
+
+  * ``distros``: Defines the list of supported distros. It can contain
```

Comment:
Is it worth noting that ``id``: Unique string beginning with the `cc_` prefix for this module.
Or maybe we just define our example in terms of the following... but then we're `import os` in each module. Though it's a paid cost anyway as we've loaded it before.

`'id':  os.path.basename(__file__).rpartition(".")[0]`